### PR TITLE
rustdesk-server doctor

### PR DESCRIPTION
**Pull request by paspo**
I added a small util to be run on a client machine to test for connectivity problems.

Example:

```bash
$ rustdesk-server doctor myrelay.rustdesk.com
Checking server:  myrelay.rustdesk.com

Found 1 IP addresses: 
 - 123.456.123.456

Checking IP address: 123.456.123.456
Is IPV4: true
Is IPV6: false
Reverse DNS lookup: 'vps444.provider.com' DOESN'T MATCH server address 'myrelay.rustdesk.com'
TCP Port 21114 (API): OK in 25 ms
TCP Port 21115 (hbbs extra port for nat test): OK in 21 ms
TCP Port 21116 (hbbs): OK in 21 ms
TCP Port 21117 (hbbr tcp): OK in 21 ms
TCP Port 21118 (hbbs websocket): OK in 24 ms
TCP Port 21119 (hbbr websocket): OK in 21 ms
```

I would like to add some more checks, like:

- ICMP ping (maybe not needed)
- key support (key required or not, check if a provided key is valid)
- client register test
- relay functionality test

As always, this is the first time I use rust, so I need vetting.

Please comment.

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -68,6 +72,79 @@ fn validate_keypair(pk: &str, sk: &str) -> ResultType<()> {
...
+    println!("Checking server:  {}\n", server_address);
+    let server_ipaddr = server_address.parse::<IpAddr>();
+    if server_ipaddr.is_err() {
+        // the passed string is not an ip address
+        let ips: Vec<std::net::IpAddr> = lookup_host(server_address).unwrap();
+        println!("Found {} IP addresses: ", ips.iter().count());
+
+        for ip in ips.iter() {
+            println!(" - {}", ip);
+        }
```
Comment (dlhxzb) [inline src/utils.rs:137]: Nitpick: `ips.iter().inspect(|ip| println!(" - {ip}");`  is more "rustacean" 🤣

Comment (paspo) [reply src/utils.rs:137]: it really looks better, but

```text
warning: unused `std::iter::Inspect` that must be used
   --> src/utils.rs:135:9
    |
135 |         ips.iter().inspect(|ip| println!(" - {ip}"));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: iterators are lazy and do nothing unless consumed
```

Which is the better approach in this case?

```rust
        let _ = ips.iter().inspect(|ip| println!(" - {ip}"));
```

maybe this?

**Comment by paspo** [comment]
> * ICMP ping (maybe not needed)
> * key support (key required or not, check if a provided key is valid)
> * client register test
> * relay functionality test

@rustdesk I need feedback on these ideas.
Does they make sense?
How can they be implemented? (I just need you to point me in the right direction, I hope I can figure out the rest)

But more important: the tests already done (DNS check, TCP ports) already covers the big part  of self-hosted user problems, so we can publish these changes and leave the ideas for the future.

If you're ok, we should probably include the rustdesk-utils binary in classic docker images also.

File: src/utils.rs (hunk 1 above)
Comment (dlhxzb) [reply src/utils.rs:137]: `inspect` -> `for_each`

Comment (paspo) [reply src/utils.rs]: fixed

**Comment by paspo** [comment]
@rustdesk can you take a look on this? I really need your feedback.

**Comment by rustdesk** [comment]
good job, you are good at rust now.

**Pull request merged**